src/tools/crawler/crawler.rs

```diff
@@ -0,0 +1,107 @@
+use std::{net::SocketAddr, str::FromStr, sync::Arc};
+
+use pea2pea::protocols::{Handshake, Reading, Writing};
+use tokio::{
+    sync::mpsc::{channel, Receiver},
+    time::{timeout, Instant},
+};
+use tracing::{info, trace, warn};
+use ziggurat_xrpl::{
+    protocol::{
+        codecs::message::{BinaryMessage, Payload},
+        proto::TmEndpoints,
+    },
+    setup::constants::CONNECTION_TIMEOUT,
+    tools::inner_node::InnerNode,
+};
+
+use crate::network::KnownNetwork;
+
+pub(super) struct Crawler {
+    node: InnerNode,
+    known_network: Arc<KnownNetwork>,
+    receiver: Receiver<(SocketAddr, BinaryMessage)>,
+}
+
+impl Crawler {
+    pub(super) async fn new() -> Self {
+        let (sender, receiver) = channel(1024);
+        let node = InnerNode::new(&Default::default(), sender).await;
+        node.enable_handshake().await;
+        node.enable_reading().await;
+        node.enable_writing().await;
+        let known_network = Arc::new(Default::default());
+        Self {
+            node,
+            known_network,
+            receiver,
+        }
+    }
+
+    pub(super) async fn start_processing(&mut self) {
+        while let Some((addr, message)) = self.receiver.recv().await {
+            if let Payload::TmEndpoints(endpoints) = message.payload {
+                self.process_endpoints_message(addr, endpoints).await;
+            }
+        }
+    }
+
+    async fn process_endpoints_message(&mut self, addr: SocketAddr, endpoints: TmEndpoints) {
+        info!(
+            "Received endpoints from {}: {:?}",
+            addr, endpoints.endpoints_v2
+        );
+        for endpoint in endpoints.endpoints_v2 {
+            match SocketAddr::from_str(&endpoint.endpoint) {
```

Comment:
at this point (after CPR tests are done), I think we can use expect here
if the node sends an invalid address, they have a bug and the crawler should just panic

Reply:
I would suggest to leave it as is for the time being - somehow I feel that panicking (and killing whole crawler) just because someone sent us data we can parse would be inappropriate here.

Reply:
At least let's change the error trace to `error` - since it's a potentially big issue on their side, no?

The whole crawler work is not really important if their nodes sometimes send invalid addresses randomly...

Reply:
Done. Although I think we have to formalize distinction between error/warn levels for the future.